theano/tensor/raw_random.py

```diff
@@ -521,12 +521,12 @@ def binomial(random_state, size=None, n=1, p=0.5, ndim=None,
     p = tensor.as_tensor_variable(p)
     ndim, size, bcast = _infer_ndim_bcast(ndim, size, n, p)
     if n.dtype == 'int64':
-        # THIS WORKS AROUND A NUMPY BUG on 32bit machine
-        # Erase when the following works on a 32bit machine:
-        # numpy.random.binomial(
-        #          n=numpy.asarray([2,3,4], dtype='int64'),
-        #          p=numpy.asarray([.1, .2, .3], dtype='float64'))
-        n = tensor.cast(n, 'int32')
+        try:
+            numpy.random.binomial(n=numpy.asarray([2, 3, 4], dtype='int64'), p=numpy.asarray([.1, .2, .3], dtype='float64'))
+        except TypeError:
+            n = tensor.cast(n, 'int32')
+    # THIS WORKS AROUND A NUMPY BUG on 32bit machine
```

Comment:
I think this section needs to be reorganized. The comment needs to go before the "try". Also, currently the code casts to int32, and then casts to int64, which defeats the purpose. What we want is for `n` to be casted in int64, except when a TypeError is raised, in which case it should be int32.
There are different ways of achieving that.